src/main/java/io/cdap/plugin/batch/source/BSONObjectToRecordTransformer.java

```diff
@@ -0,0 +1,201 @@
+      case DOUBLE:
+        ensureTypeValid(fieldName, object, Double.class);
+        return object;
+      case BYTES:
+        ensureTypeValid(fieldName, object, ObjectId.class, byte[].class);
+        if (object instanceof ObjectId) {
+          // ObjectId can be read as a bytes. Exactly 12 bytes are required
+          ByteBuffer buffer = ByteBuffer.allocate(12);
+          ObjectId objectId = (ObjectId) object;
+          objectId.putToByteBuffer(buffer);
+          return buffer;
+        }
+        return object;
+      case LONG:
+        ensureTypeValid(fieldName, object, Long.class);
+        return object;
+      case STRING:
+        ensureTypeValid(fieldName, object, ObjectId.class, String.class);
+        if (object instanceof ObjectId) {
+          // ObjectId can be read as a string
+          return object.toString();
+        }
+        return object;
+      case MAP:
+        ensureTypeValid(fieldName, object, BSONObject.class);
+        return convertMap(fieldName, (BSONObject) object, schema);
+      case ARRAY:
+        ensureTypeValid(fieldName, object, BasicBSONList.class);
+        return convertArray(fieldName, (BasicBSONList) object, schema);
+      case RECORD:
+        ensureTypeValid(fieldName, object, BSONObject.class);
+        return convertRecord(fieldName, (BSONObject) object, schema);
+      default:
+        throw new UnexpectedFormatException("Field type '" + fieldType + "' is not supported.");
+    }
+  }
+
+  private Object convertMap(String fieldName, BSONObject bsonObject, Schema schema) {
+    Schema valueSchema = schema.getMapSchema().getValue().isNullable()
+      ? schema.getMapSchema().getValue().getNonNullable() : schema.getMapSchema().getValue();
+    Map<String, Object> extracted = new HashMap<>();
+    for (String key : bsonObject.keySet()) {
+      Object value = bsonObject.get(key);
+      extracted.put(key, extractValue(fieldName, value, valueSchema));
+    }
+    return extracted;
+  }
+
+  private Object convertArray(String fieldName, BasicBSONList bsonList, Schema schema) {
+    List<Object> values = Lists.newArrayListWithCapacity(bsonList.size());
+    Schema componentSchema = schema.getComponentSchema().isNullable() ? schema.getComponentSchema().getNonNullable()
+      : schema.getComponentSchema();
+    for (Object obj : bsonList) {
+      values.add(extractValue(fieldName, obj, componentSchema));
+    }
+    return values;
+  }
+
+  private void ensureTypeValid(String fieldName, Object value, Class... expectedTypes) {
+    for (Class expectedType : expectedTypes) {
+      if (expectedType.isInstance(value)) {
+        return;
+      }
+    }
+
+    String expectedTypeNames = Stream.of(expectedTypes)
+      .map(Class::getName)
+      .collect(Collectors.joining(", "));
+    throw new UnexpectedFormatException(
+      String.format("Document field '%s' is expected to be of type '%s', but found a '%s'.", fieldName,
+                    expectedTypeNames, value.getClass().getName()));
```

Comment:
will be more readable to use the simple class name rather than a full java class name.

Reply:
Done.